google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb

```diff
@@ -0,0 +1,214 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "google/cloud/pubsub/service"
+require "google/cloud/errors"
+require "monitor"
+require "forwardable"
+require "concurrent"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # # Subscriber
+      #
+      class Subscriber
+        include MonitorMixin
+
+        ##
+        # @private Implementation attributes.
+        attr_reader :request_queue, :output_enum, :thread_pool, :service
+
+        ##
+        # Subscriber attributes.
+        attr_reader :callback, :subscription, :deadline, :threads
+
+        ##
+        # @private Create an empty {Subscriber} object.
+        def initialize callback, subscription, deadline, threads, service
+          @request_queue = nil
+          @output_enum = nil
+          @callback = callback
+          @subscription = subscription
+          @deadline = deadline
+          @threads = threads || [2, Concurrent.processor_count * 2].max
+          @service = service
+
+          super() # to init MonitorMixin
+        end
+
+        def start
+          synchronize do
+            return if @request_queue
+
+            @request_queue = EnumeratorQueue.new self
+            @thread_pool = Concurrent::FixedThreadPool.new threads
+            start_streaming!
+          end
+
+          true
+        end
+
+        def stop
+          synchronize do
+            return if @request_queue.nil?
+            @request_queue.push self
+          end
+
+          true
+        end
+
+        def wait!
+          synchronize do
+            return if @background_thread.nil?
+            @background_thread.join
+          end
+
+          return true if @thread_pool.shutdown?
+          @thread_pool.shutdown
+          @thread_pool.wait_for_termination
+
+          true
+        end
+
+        ##
+        # @private
+        def acknowledge *messages
+          ack_ids = coerce_ack_ids messages
+          return true if ack_ids.empty?
+
+          ack_request = Google::Pubsub::V1::StreamingPullRequest.new
+          ack_request.ack_ids += ack_ids
+
+          synchronize do
+            @request_queue.push ack_request
+          end
+
+          true
+        end
+
+        ##
+        # @private
+        def delay new_deadline, *messages
```

Comment:
I think I would prefer this be named `modify_ack_deadline` to match the API (and the method name in other languages). It actually took me a moment to figure out what this method did.

I would also note that the method is _not_ a delay. You can call this method to make the deadline _shorter_. For example, if you call:

```ruby
delay 20 message
delay 1 message
```

The new deadline will be 1 second from when the latter call went in, not 21 seconds.
(Gotcha: Unless those calls end up getting processed out of order, but I am not too worried about this as this is a private method and calling it in such a way is an obvious anti-pattern.)

The verb `delay` sounds to me like "add N seconds to this timeout", when the behavior is actually "set the timeout to N seconds".